README.md
# deploy-functions

[![Build Status](https://drone.osshelp.ru/api/badges/ansible/deploy-functions/status.svg)](https://drone.osshelp.ru/ansible/deploy-functions)

## About

This library is used for customization in the deploying process. It is used with default-setup script. And default-setup is run by cloud-init.

Supported software:

* Netdata
* Monit
* Postfix
* Ssmtp
* Nginx
* PHP-FPM
* Apache
* Cron
* SSH
* MySQL
* Consul
* Promtail

### How to use it

The library must be included in the default-setup script. How to use the library:

1. Place deploy-functions.sh library in the `/usr/local/include/osshelp/` folder
1. Place default-setup script in the `/usr/local/osshelp/` folder with default-setup.local (if you need an additional customization)
1. Install yq and jq binaries
1. Add necessary profiles to container yaml config which is used by lxhelper
1. Deploy container with lxhelper and make sure that the default-setup works as you expected

In the examples directory you can find a template of default-setup.local and available profiles usage examples.

There’re install/update scripts in the repository. Command for installation (deploy-function.sh, default-setup, yq and jq):

```shell
curl -s https://oss.help/scripts/backup/backup-functions/install.sh | bash
```

Or you can use this [Ansible role](https://github.com/OSSHelp/ansible-deploy-functions) to install it.

### Useful

As a library, deploy-functions is often used with [LXHelper](https://github.com/OSSHelp/lxhelper), but it's still possible to use it without LXHelper.

## FAQ

### Log

You can find log in `/var/log/default-setup.log`.

### Debug mode

If you want to run default-setup in debug mode, you need to:

* Export INSTANCE_ID variable in the container
* Run default-setup with `bash -x` command

Example:

```bash
export INSTANCE_ID=container_name
bash -x /usr/local/osshelp/default-setup server_name
```

## Author

OSSHelp Team, see <https://oss.help>


CHANGELOG.md
# Changelog

## v1.25

* fixed a bug with source label generic value not being replaced with INSTANCE_ID value in Promtail cfg

## v1.24

* Fixed a bug with source editing for expression in match, etc
* Added source_exclude variable to promtail profile

## v1.23

* Fixed a bug with `source` label edition by default-setup

## v1.22

* Server name is now used as the default value for Promtail instance labels
* Updated Promtail's default config name

## v1.20

* Container name is now used as the default value for Promtail instance/source labels

## v1.19

* Auto replacement of instance/source labels for Promtail
* Improved Promtail limits_config section generation (any future param changes support)

## v1.18

* Added Promtail profiles support

## v1.13

* Added default logs generation opportunity for Focal and Bionic (as it was made for Xenial and Trusty)

## v1.12

* Added Consul profiles support

## v1.11

* Added profiles names support for default_setup.local

## v1.10

* Fixed a bug when default-setup stops if profiles is absent in the container
* Fixed bugs in the functions that customize services (Nginx, Monit, etc)

## v1.09

* Fixed bug when default-setup sometimes stops if there was null value in profiles

## v1.08

* Nginx.stat-server-name key was renamed to nginx.stat-servername
* Added MySQL configuration files management
* Added system users management (lock/unlock)
* Updated setup_postfix function. Added header customization opportunity (for Cron emails).
* Updated setup_monit function. Now it doesn't delete files from conf.d but are moved to conf-disabled. It deletes only symlinks from conf-enabled. Fixed mail format in the main configuration file.
* Updated a list of folders for generation in `/var/log` (if they are absent)
* Added the package list checking. If a package is installed the library adds a folder into `/var/log`. Supported packages:
  * nginx
  * nginx-light
  * nginx-full
  * nginx-extras
  * php5-fpm
  * php-fpm5.5
  * php-fpm5.6
  * php-fpm7.0
  * php-fpm7.1
  * redis-server
  * proftpd-basic
  * lxd
  * MySQL
* Fixed a bug in the generation process of Monit main configuration file (related with monit.start-delay key)
* Fixed behavior of some functions. Now they don't try to change files if they are absent. This fix reduced log output.
* Fixed functions call order in default-setup

## v1.06

* Changed log path by /var/log/default-setup.log
* Added hlreports parameters management
* Fixed the type of /usr/local/osshelp/default-setup.local check
* Deleted duplicated functions (make_dir/make_file)
* Fixed bug with missing separator in `/etc/netdata/stream.conf` configuration process
* Added highload configuration (it works when Monit is present)
* Updated setup_phpfpm function (enable/disable ini files)


deploy-functions/install.sh
#!/bin/bash
# shellcheck disable=SC2015

umask 0077
export LANG=C
export LC_ALL=C

df_ver=1.25
silent=no
tmpdir="${TEMP:=/tmp}"

index_url="${REMOTE_URI:-https://oss.help/scripts/lxc/deploy-functions/.list}"
list_name=$(basename "${index_url}")
script_name="default-setup"
script_path="/usr/local/osshelp"
lib_name="deploy-functions.sh"
lib_path="/usr/local/include/osshelp"
yq_bin='yq'
yq_bin_path='/usr/local/bin'
shacmd=$(command -v sha256sum || command -v gsha256sum 2>/dev/null)
err=0

function show_notice() { test "${silent}" != "yes" && echo -e "[NOTICE] ${*}"; return 0; }
function show_error() { echo -e "[ERROR] ${*}" >&2; err=1; return 1; }

function fetch_files() {
  cd "${1}" && {
    {
      wget -q -P "${1}" "${index_url}" && \
        wget -q -i "${1}/${list_name}" -P "${1}"
        test -f  SHA256SUMS || show_error "File SHA256SUMS does not found."
    } && {
      "${shacmd}" -c --status SHA256SUMS 2> /dev/null || {
        show_error "Something went wrong, checksums of downloaded files mismatch."
        "${shacmd}" -c "${1}/SHA256SUMS"
        return 1
      }
    }
  }
}

function install_files() {
        test -d "${script_path}" || mkdir "${script_path}"
        cd "${script_path}" && \
        mv "${tmp_dir}/${script_name}" "${script_path}/${script_name}" && \
        chmod 700 "${script_path}/${script_name}"

        test -d "${lib_path}" || mkdir "${lib_path}"
        cd "${lib_path}" && \
        mv "${tmp_dir}/${lib_name}" "${lib_path}/${lib_name}" && \
        chmod 600 "${lib_path}/${lib_name}"

       test -x "${script_path}/${script_name}" || show_error "${script_name} hasn't been installed."
       test -f "${lib_path}/${lib_name}"|| show_error "${lib_name} hasn't been installed."
}

function install_jq() {
        command -v apt-get > /dev/null 2>&1 && {
            apt-get -qq update && apt-get -qqy install jq
        }
        command -v yum > /dev/null 2>&1 && yum -qy install jq
        command -v jq > /dev/null 2>&1 && show_notice "Jq ($(jq --version)) has been installed."
        command -v jq > /dev/null 2>&1 || show_error "Jq hasn't been installed."
}

test "$(id -u)" != "0" && { show_error "Sorry, but you must run this script as root."; exit 1; }

tmp_dir="${tmpdir}/deploy-functions.$$"
mkdir -p "${tmp_dir}" && \
fetch_files "${tmp_dir}" && \
install_files "${tmp_dir}" && {
        test -x "${yq_bin_path}/${yq_bin}" || curl -s https://oss.help/scripts/tools/yq/install.sh | bash
        install_jq
        show_notice "Library ${lib_name} (v${df_ver}) was installed to ${lib_path}."
        show_notice "Script ${script_name} was installed to ${script_path}."
}

test -d "${tmp_dir}" && rm -rf "${tmp_dir}"
exit "${err}"


deploy-functions/update.sh
#!/bin/bash
# shellcheck disable=SC2015

umask 0077
export LANG=C
export LC_ALL=C

df_ver=1.25
silent=no
tmpdir="${TEMP:=/tmp}"

index_url="${REMOTE_URI:-https://oss.help/scripts/lxc/deploy-functions/.list}"
list_name=$(basename "${index_url}")
script_name="default-setup"
script_path="/usr/local/osshelp"
lib_name="deploy-functions.sh"
lib_path="/usr/local/include/osshelp"
yq_bin='yq'
yq_bin_path='/usr/local/bin'
shacmd=$(command -v sha256sum || command -v gsha256sum 2>/dev/null)
err=0

function show_notice() { test "${silent}" != "yes" && echo -e "[NOTICE] ${*}"; return 0; }
function show_error() { echo -e "[ERROR] ${*}" >&2; err=1; return 1; }

function fetch_files() {
  cd "${1}" && {
    {
      wget -q -P "${1}" "${index_url}" && \
        wget -q -i "${1}/${list_name}" -P "${1}"
        test -f  SHA256SUMS || show_error "File SHA256SUMS does not found."
    } && {
      "${shacmd}" -c --status SHA256SUMS 2> /dev/null || {
        show_error "Something went wrong, checksums of downloaded files mismatch."
        "${shacmd}" -c "${1}/SHA256SUMS"
        return 1
      }
    }
  }
}

function install_files() {
        test -d "${script_path}" || mkdir "${script_path}"
        cd "${script_path}" && \
        mv "${tmp_dir}/${script_name}" "${script_path}/${script_name}" && \
        chmod 700 "${script_path}/${script_name}"

        test -d "${lib_path}" || mkdir "${lib_path}"
        cd "${lib_path}" && \
        mv "${tmp_dir}/${lib_name}" "${lib_path}/${lib_name}" && \
        chmod 600 "${lib_path}/${lib_name}"

        test -x "${script_path}/${script_name}" || show_error "${script_name} hasn't been installed."
        test -f "${lib_path}/${lib_name}"|| show_error "${lib_name} hasn't been installed."

}

function install_jq() {
        command -v apt-get > /dev/null 2>&1 && {
            apt-get -qq update && apt-get -qqy install jq
        }
        command -v yum > /dev/null 2>&1 && yum -qy install jq
        command -v jq > /dev/null 2>&1 && show_notice "Jq ($(jq --version)) has been installed."
        command -v jq > /dev/null 2>&1 || show_error "Jq hasn't been installed."
}

test "$(id -u)" != "0" && { show_error "Sorry, but you must run this script as root."; exit 1; }

tmp_dir="${tmpdir}/deploy-functions.$$"
mkdir -p "${tmp_dir}" && \
fetch_files "${tmp_dir}" && \
install_files "${tmp_dir}" && {
        test -x "${yq_bin_path}/${yq_bin}" && curl -s https://oss.help/scripts/tools/yq/update.sh | bash
        test -x "${yq_bin_path}/${yq_bin}" || curl -s https://oss.help/scripts/tools/yq/install.sh | bash
        install_jq
        show_notice "Library ${lib_name} was updated up to version v${df_ver}."
        show_notice "Script ${script_name} was updated."
}

test -d "${tmp_dir}" && rm -rf "${tmp_dir}"
exit "${err}"
